rectools/models/nn/transformers/lightning.py

```diff
@@ -149,15 +151,17 @@ def training_step(self, batch: tp.Dict[str, torch.Tensor], batch_idx: int) -> to
         """Training step."""
         x, y, w = batch["x"], batch["y"], batch["yw"]
         if self.loss == "softmax":
-            logits = self._get_full_catalog_logits(x)
+            _, _, logits = self.torch_model(sessions=x)
```

Comment:
This is just a suggestion. But to clean up all of our modifications code we can do the following:

- create a property `requires_negatives`. True for softmax, False for negative losses, None for unknown losses.
- Get logits depending on the property with common code for all losses.
- Then call just `if` for each lossname and calculated.

Or even cleaner: reuse TorchRanker logic `_get_scorer` logic from here: https://github.com/MobileTeleSystems/RecTools/blob/c5dc5f793e3727aade308fcdb3563f0e4c3d5fb7/rectools/models/rank/rank_torch.py#L178
We can create `_get_loss_calculator` which returns a flag for requiring negatives and a function to call to calculate loss.

What do you think?

Reply:
- About reusing TorchRanker:
We should put the following methods `_calc_softmax_loss`, `_calc_bce_loss`, `_calc_gbce_loss` + `_get_reduced_overconfidence_logits` and `_calc_custom_loss_outputs` in `TransformerLightningModuleBase` so that you don't create another `__init__` in `TransformerLightningModule`. But we can't tell the difference between `bce` and `gbce` without an additional if, since the negatives are fed to the input of the gbce loss.

So I think we won't get away from if in these methods, so I consider the first one to be a cleaner option

Reply:
I selected option with properties `requires_negatives`

Reply:
Seems good to me